### PR TITLE
feat(routing): custom header-triggered routing tiers

### PR DESCRIPTION
**Pull request by brunobuddy**
## Summary

Adds a **third routing axis** that overrides specificity and complexity: a user-configurable tier keyed to an incoming HTTP header key+value pair (e.g. `x-manifest-tier: premium`). When a request arrives with a matching header, Manifest forwards directly to the tier's configured model, stamps the message with the tier's id/name/color, and labels the routing reason `header-match`.

Lets users point any SDK's identifier (whether it's `x-manifest-tier`, `x-stainless-package-version`, `x-title`, or a custom app header) at a specific model. The create modal autocompletes the header key from Manifest's own `agent_messages.request_headers` data (shipped in #1637) — so the user picks from headers they've actually been receiving, grouped by the SDK that sent them, with the 3 most common values surfaced inline.

### Backend (`packages/backend/`)
- New `header_tiers` table + 3 denormalized columns on `agent_messages` (`header_tier_{id,name,color}`). Denormalized on purpose: historical rows keep the label they matched at the time, and Messages-log reads stay a flat SELECT.
- `HeaderTierService` — CRUD + reorder + override/fallbacks. Validates name uniqueness (case-insensitive), header-key deny-list (`authorization`, `cookie`, `set-cookie`, `proxy-authorization`, `x-api-key`), key regex, rule uniqueness, color enum, length caps.
- `HeaderTierController` — 9 REST endpoints under `/routing/:agentName/header-tiers`.
- `SeenHeadersService` — Postgres aggregate over `request_headers::jsonb` with per-key top-3 values and SDK attribution (from `caller_attribution.sdk`). Tenant-scoped, 5-minute TTL cache, 50-key / 7-day window.
- `resolveHeaderTier()` runs **first** in `ResolveService.resolve()` — before specificity and complexity. First-match-wins in `sort_order`; tiers without a configured model fall through so complexity routing still works.
- `RoutingMeta` grows `header_tier_{id,name,color}` fields; threaded through every recorder write path (5 sites, mirroring #1637's `requestHeaders` pattern), plus the `proxy-response-handler` helpers and `handleProxyError`.
- `selectMessageRowColumns()` + `MessageDetailResponse` expose the 3 new fields so the Messages list + detail view can render them.

### Frontend (`packages/frontend/`)
- New `RoutingHeaderTiersSection` mounted on `Routing.tsx` **above** the generalist tiers — visual order matches execution order.
- `HeaderTierCard` — compact horizontal card with ordinal chip (`#N`), colored dot, tier name, mono rule chip (`key: value`), model slot (reuses `ModelPickerModal`), Reset, and kebab menu with Delete.
- `CreateHeaderTierModal` — single-step form. Inputs: name → header key (combobox) → header value (combobox filtered by key) → 6-swatch color picker (slate / indigo / teal / amber / rose / violet).
- `HeaderComboBox` — shared typeahead used in the modal. Shows `seen N×` counts + up to 3 example values. Keyboard nav (↑/↓ Enter Esc). Free-form override pinned to the footer when the user's input doesn't match any known header.
- Message badge: new `tier-badge--custom` branch in `message-table-cells.tsx`, colored from `tier-color--{slate|indigo|teal|amber|rose|violet}`. Priority: custom → specificity → tier.
- `MessageDetails` `Routing` meta shows the tier name when `header-match` fires; falls back to the existing specificity / tier label otherwise.

## Test plan

- [x] Backend unit tests: 4050/4050 pass
  - `HeaderTierService`: 30 tests covering CRUD, validation (deny-list, uniqueness, regex, caps, color enum), reorder edge cases, override/fallback paths
  - `SeenHeadersService`: 5 tests covering query projection, tenant vs user scoping, agent filter, 5-min cache, LRU eviction
  - `ResolveService`: 8 new header-tier cases (match, array-header match, empty-model fallthrough, key missing, value mismatch, no-headers skip, first-match-wins, model-prefix provider inference)
  - Recorder / response-handler / proxy.controller specs all updated for the new fields
- [x] Frontend unit tests: 2269/2269 pass (no existing tests broken)
- [x] Typecheck clean in both workspaces
- [x] Smoke test on fresh Postgres: migration creates `header_tiers` + 3 `agent_messages` columns; inserting a tier + hitting `POST /v1/chat/completions` with the matching header routes through `resolveHeaderTier` (friendly "no OpenAI key" response confirms the tier's configured model was picked, not the default seed routing).
- [ ] Manual UI walk: create tier in Routing page, verify it persists across reload, send matching request, see colored badge on the message row.

---
## Summary by cubic
Adds custom header-triggered routing tiers that send matching requests straight to a chosen model. Includes a snippet modal, manage modal with enable/disable, and preserves `routing_reason` on all fallback and error paths; UI now says “Complexity tiers” and “Specificity tiers”.

- **New Features**
  - Routing: header tiers resolve before specificity/complexity; first-match by `sort_order`; disabled tiers are ignored; routing reason `header-match`; proxy forwards request headers into resolve.
  - API/Data: manage at `/api/v1/routing/:agentName/header-tiers` (CRUD, reorder, override, fallbacks, toggle); autocomplete at `/api/v1/routing/:agentName/seen-headers` (supports `scope=all`); `agent_messages` and resolve responses include `header_tier_{id,name,color}`.
  - UI: “Custom tiers” section after Complexity → Specificity; compact cards with model override/reset and a gear to open the snippet modal; manage modal with enable/disable, edit, and delete; 24 badge colors; shared `.modal-back-btn`; Messages show a colored badge and Details show the tier name.
  - SDK snippets: `FrameworkSnippets` accepts `customHeaders`; headers weave into OpenAI (`default_headers`/`defaultHeaders`), Vercel AI (`headers`), LangChain (`default_headers`), and cURL (`-H`); connection details list each header.

- **Bug Fixes**
  - Auditing: `routing_reason` persisted on all fallback rows and no-fallback error paths; `meta.reason` forwarded through response-handler paths.
  - Validation/Robustness: blocks sensitive keys; enforces unique names/rules and header-key regex; rejects quotes/backslashes (now blocks single quotes too); orphaned override models fall through; reorder validates type/length/membership and duplicates; controller validates `FallbacksBody.models`; seen-headers `top_values` ranked by frequency.
  - Copy/UX: rename “Capability tiers” to “Specificity tiers”; dark-mode icons fixed; consistent back buttons; header-tier modal adds back button and delete confirm.
  - Tests: expanded coverage for header-tier UI/API including manage modal (open/close, overlay/escape, toggle, edit, back, delete, create-new), `toggleHeaderTier`, and modal back/delete-confirm; removed a dead `handleReset` path; provider flows updated (e.g., ModelCell badge).

<sup>Written for commit 95cf20a4d00424306364d606cce856dbe6cfd1cb. Summary will update on new commits.</sup>

### CODE REVIEW

File: packages/backend/src/routing/resolve/resolve.service.ts [hunk 1]
```diff
@@ -139,6 +149,50 @@ export class ResolveService {
...
+    if (!match) return null;
+
+    if (!match.override_model) {
+      this.logger.debug(
+        `Header tier "${match.name}" matched but has no model configured — falling through`,
+      );
+      return null;
+    }
+
+    const provider = await this.resolveProvider(
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/routing/resolve/resolve.service.ts:180]: Fixed in f762b267.

File: packages/backend/src/routing/header-tiers/header-tier.controller.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  @Post(':agentName/header-tiers/reorder')
+  async reorder(
+    @CurrentUser() user: AuthUser,
+    @Param('agentName') agentName: string,
+    @Body() body: ReorderBody,
+  ) {
+    const agent = await this.resolveAgentService.resolve(user.id, agentName);
+    await this.headerTierService.reorder(agent.id, body.ids);
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/routing/header-tiers/header-tier.controller.ts:118]: Fixed in f762b267.

File: packages/frontend/src/components/HeaderTierCard.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+          >
+            ⋯
+          </button>
+          <Show when={kebabOpen()}>
+            <div class="header-tier-card__menu" role="menu">
+              <button
+                type="button"
+                role="menuitem"
+                class="header-tier-card__menu-item header-tier-card__menu-item--danger"
+                onMouseDown={(e) => {
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/frontend/src/components/HeaderTierCard.tsx]: Fixed in f762b267.

File: packages/backend/src/routing/header-tiers/header-tier.service.ts [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+  async delete(agentId: string, id: string): Promise<void> {
+    const row = await this.findOrThrow(agentId, id);
+    await this.repo.delete({ id: row.id });
+    this.routingCache.invalidateAgent(agentId);
+  }
+
+  async reorder(agentId: string, ids: string[]): Promise<void> {
+    const rows = await this.repo.find({ where: { agent_id: agentId } });
+    const byId = new Map(rows.map((r) => [r.id, r]));
+    if (ids.length !== rows.length || !ids.every((id) => byId.has(id))) {
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/routing/header-tiers/header-tier.service.ts]: Fixed in f762b267.

File: packages/backend/src/routing/header-tiers/seen-headers.service.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    if (agentName) {
+      params.push(agentName);
+      filters.push(`at.agent_name = $${params.length}`);
+    }
+
+    const sql = `
+      SELECT
+        key,
+        COUNT(*)::int AS count,
+        (array_agg(DISTINCT value))[1:${MAX_VALUES_PER_KEY}] AS top_values,
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/routing/header-tiers/seen-headers.service.ts]: Fixed in f762b267.

File: packages/backend/src/entities/header-tier.entity.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import { Entity, Column, PrimaryColumn, Index } from 'typeorm';
+import { timestampType, timestampDefault } from '../common/utils/sql-dialect';
+import type { TierColor } from 'manifest-shared';
+
+@Entity('header_tiers')
+@Index(['agent_id', 'sort_order'])
+@Index(['agent_id', 'name'], { unique: true })
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/entities/header-tier.entity.ts]: Fixed in f762b267.

File: packages/backend/src/routing/proxy/__tests__/proxy.controller.spec.ts [hunk 7]
```diff
@@ -823,6 +823,7 @@ describe('ProxyController', () => {
       tenantId: 'tenant-1',
       agentName: 'test-agent',
       signal: expect.any(AbortSignal),
+      headers: expect.any(Object),
     });
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/routing/proxy/__tests__/proxy.controller.spec.ts]: Fixed in f762b267.

File: packages/frontend/src/components/HeaderTierCard.tsx [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+          >
+            ⋯
+          </button>
+          <Show when={kebabOpen()}>
+            <div
+              class="header-tier-card__menu"
+              role="menu"
+              // Close the menu when focus leaves it so blur doesn't need to
+              // race with the menu item's own click handler.
+              onFocusOut={(e) => {
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/frontend/src/components/HeaderTierCard.tsx]: Fixed in b4b7a638.

**Comment by brunobuddy** [comment]
@cubic-dev-ai review

**Comment by brunobuddy** [comment]
@cubic-dev-ai review (re-requesting after merge commit 1bc7776a)

File: packages/backend/src/routing/header-tiers/header-tier.controller.ts [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+  @Put(':agentName/header-tiers/:id/fallbacks')
+  async setFallbacks(
+    @CurrentUser() user: AuthUser,
+    @Param('agentName') agentName: string,
+    @Param('id') id: string,
+    @Body() body: FallbacksBody,
+  ) {
+    const agent = await this.resolveAgentService.resolve(user.id, agentName);
+    return this.headerTierService.setFallbacks(agent.id, id, body.models);
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/routing/header-tiers/header-tier.controller.ts:158]: Fixed in f7712367

File: packages/frontend/src/components/HeaderTierCard.tsx [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+                            'justify-content': 'center',
+                            width: '18px',
+                            height: '18px',
+                            'border-radius': '50%',
+                            'font-size': '10px',
+                            'font-weight': '600',
+                          }}
+                        >
+                          <Show when={cp} fallback={(cp?.name ?? 'C').charAt(0).toUpperCase()}>
+                            {customProviderLogo(cp!.name, 14)}
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/frontend/src/components/HeaderTierCard.tsx]: Fixed in f7712367

File: packages/frontend/src/services/framework-snippets.ts [hunk 11]
```diff
@@ -169,14 +217,23 @@ openclaw config set agents.defaults.model.primary ${model}
...
 
-export function getCurlSnippet(baseUrl: string, apiKey: string): Snippet[] {
+export function getCurlSnippet(
+  baseUrl: string,
+  apiKey: string,
+  customHeaders?: CustomHeaders,
+): Snippet[] {
+  const extraHeaders = customHeaders
+    ? Object.entries(customHeaders)
+        .map(([k, v]) => `  -H "${k}: ${v}" \\\n`)
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/frontend/src/services/framework-snippets.ts]: Fixed in f7712367

File: packages/frontend/src/services/framework-snippets.ts [hunk 12]
```diff
@@ -69,7 +69,44 @@ export function storeFramework(id: FrameworkId): void {
...
+
+/** `{"x-key": "value", "x-other": "v"}` rendered for the target language. */
+function renderHeadersDict(headers: CustomHeaders | undefined, lang: 'py' | 'ts'): string | null {
+  if (!headers) return null;
+  const entries = Object.entries(headers);
+  if (entries.length === 0) return null;
+  if (lang === 'py') {
+    return `{${entries.map(([k, v]) => `"${k}": "${v}"`).join(', ')}}`;
+  }
+  return `{ ${entries.map(([k, v]) => `"${k}": "${v}"`).join(', ')} }`;
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/frontend/src/services/framework-snippets.ts]: Fixed in f7712367

File: packages/backend/src/routing/proxy/proxy-response-handler.ts [hunk 13]
```diff
@@ -85,9 +85,13 @@ export async function handleProviderError(
       fallbackFromModel: meta.fallbackFromModel,
       fallbackIndex: meta.fallbackIndex,
       authType: meta.auth_type,
+      reason: meta.reason,
       specificityCategory: meta.specificity_category,
       callerAttribution,
       requestHeaders,
+      headerTierId: meta.header_tier_id,
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/routing/proxy/proxy-response-handler.ts:92]: Fixed in f7712367
